### PR TITLE
Enhancement/9056 fetch all taxonomies

### PR DESCRIPTION
**Pull request by BrookeGraham**
## Context

fetch all pages, one by one, like Gutenberg does.

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1. add over 100 categories
2. load page and look at network for more then one request for categories

### Does this PR have a security-related impact?
no

### Does this PR change what data or activity we track or use?
no

### Does this PR have a legal-related impact?
no

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #
9056

### CODE REVIEW

File: packages/story-editor/src/app/taxonomy/taxonomyProvider.js [hunk 1]
```diff
@@ -191,11 +204,29 @@ function TaxonomyProvider(props) {
...
 
+  // Fetch hierarchical taxonomies on mount
+  useEffect(() => {
+    // only fetch when `shouldRefetchCategories` is true
+    if (shouldRefetchCategories && taxonomies?.length) {
+      const hierarchicalTaxonomies = taxonomies.filter(
+        (taxonomy) => taxonomy.hierarchical
+      );
+      hierarchicalTaxonomies.forEach((taxonomy) =>
+        addSearchResultsToCache(taxonomy, { perPage: -1 })
```
Comment (maxyinger) [inline packages/story-editor/src/app/taxonomy/taxonomyProvider.js:223]: Does `per_page: -1` fetch all items? I'm not seeing this being a feature of the wp rest api in the docs:
https://developer.wordpress.org/rest-api/using-the-rest-api/pagination/

I think we may have to fetch all pages like:
```
promiseAll([
  fetchPage1(),
  fetchPage2(),
  ...etc
])
```
we should be able to get the total number of pages off of the `X-WP-TotalPages` header attribute of the response.

Bummer. it looks like the old code we had in the dashboard that did something similar got deleted in the past week or so. let me see if I can dig up an old commit where we can reference it

**Review by maxyinger — CHANGES_REQUESTED** [review]

File: packages/story-editor/src/app/taxonomy/taxonomyProvider.js (hunk 1 above)
Comment (maxyinger) [reply packages/story-editor/src/app/taxonomy/taxonomyProvider.js:223]: Here's that old function for reference that fetched all pages. will have to adapt it a bit, but should be a good reference point:
https://github.com/google/web-stories-wp/pull/9065/files#r712270861

Comment (samwhale) [reply packages/story-editor/src/app/taxonomy/taxonomyProvider.js:223]: Pascal added this to the ticket: https://github.com/google/web-stories-wp/issues/9056#issuecomment-922013624

>Pass -1 as the number to fetch all pages, one by one, like Gutenberg does.

So I guess we'd only need to make one call?

**Comment by samwhale** [comment]
@googlebot I consent.

File: packages/story-editor/src/app/taxonomy/taxonomyProvider.js (hunk 1 above)
Comment (BrittanyIRL) [reply packages/story-editor/src/app/taxonomy/taxonomyProvider.js:223]: yup! passing `-1` should return us all of the pages - you can see this in action if you have a bunch of categories on a post, you'll see multiple calls get made to grab all the pages. I haven't tested this one yet but I'm assuming it's the same way. 

This is also why I went ahead and deleted that old util. :)

Comment (maxyinger) [reply packages/story-editor/src/app/taxonomy/taxonomyProvider.js:223]: oh that's awesome

**Pull request closed without merging**